File: ex1008.java

// Definition for a binary tree node.
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}


class Solution {
    public void insert(TreeNode root, int value) {
        TreeNode p = root, q = root;
        TreeNode newNode = new TreeNode(value);
        
        while (p != null) {
            if (value < p.val)    {q = p;    p = p.left;}
            else if (value > p.val)    {q = p;    p = p.right;}
        }
        
        if (value < q.val)    q.left = newNode;
        else if (value > q.val)    q.right = newNode;
    }
    
    
    public TreeNode bstFromPreorder(int[] preorder) {
        if (preorder.length == 0)    return null;
        
        
        TreeNode root = new TreeNode(preorder[0]);
        
        for (int i = 1; i < preorder.length; i++) {
            insert(root, preorder[i]);
        }
        
        return root;
    }
}

File: code/ex1008.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        return helper(preorder, 0, preorder.size() - 1);
    }
    
private:
    TreeNode* helper(vector<int>& nodes, int lo, int hi) {
        if (lo > hi)    return NULL;
        
        int r = nodes[lo];
        TreeNode* root = new TreeNode(r);
        
        // Get split point
        int i = lo;
        while (i <= hi) {
            if (i == hi || nodes[i + 1] > r) {
                root->left = helper(nodes, lo + 1, i);
                root->right = helper(nodes, i + 1, hi);
                break;
            }
            i++;
        }
        
        return root;
    }
};